Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,220 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE MultiWayIf #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+#include "MachDeps.h"
+
+-- This file is included by "Data.ByteString.ReadInt",
+-- "Data.ByteString.ReadInt64", and "Data.ByteString.Lazy.ReadInt64", after
+-- defining "BYTESTRING_STRICT" or "API_BITS_64" or both.  The modules are
+-- largely identical, except for the choice of ByteString type, and the
+-- the use of native words, or explicit 64-bits words for the accumulator.
+
+#ifndef API_BITS_64
+
+# ifdef BYTESTRING_STRICT
+module Data.ByteString.ReadInt
+# else
+module Data.ByteString.Lazy.ReadInt
+# endif
+    ( readInt
+    , readInt8
+    , readInt16
+    , readInt32
+    ) where
+
+#else
+
+# ifdef BYTESTRING_STRICT
+module Data.ByteString.ReadInt64
+# else
+module Data.ByteString.Lazy.ReadInt64
+# endif
+    ( readInt64 ) where
+
+#endif
+
+import qualified Data.ByteString.Internal as BI
+import qualified Data.ByteString.Unsafe as BU
+#ifdef BYTESTRING_STRICT
+import Data.ByteString as B
+#else
+import Data.Bits ((.|.))
+import Data.ByteString.Lazy as B
+import Data.ByteString.Lazy.Internal
+#endif
+import Data.Int
+import Data.Word
+import Foreign.Ptr (minusPtr, plusPtr)
+import Foreign.Storable (Storable(..))
+
+----- Public API
+
+#ifndef API_BITS_64
+type Accum = Word
+
+-- | Try to read a signed 'Int' value from the 'ByteString', returning
+-- @Just (val, str)@ on success, where @val@ is the value read and @str@ is the
+-- rest of the input string.  If the sequence of digits decodes to a value
+-- larger than can be represented by an 'Int', the returned value will be
+-- 'Nothing'.
+--
+-- 'readInt' does not ignore leading whitespace, the value must start
+-- immediately at the beginning of the input string.
+--
+-- ==== __Examples__
+-- >>> readInt "-1729 sum of cubes"
+-- Just (-1729," sum of cubes")
+-- >>> readInt "+1: readInt also accepts a leading '+'"
+-- Just (1, ": readInt also accepts a leading '+'")
+-- >>> readInt "not a decimal number")
+-- Nothing
+-- >>> readInt "12345678901234567890 overflows maxBound")
+-- Nothing
+-- >>> readInt "-12345678901234567890 underflows minBound")
+-- Nothing
+--
+readInt :: ByteString -> Maybe (Int, ByteString)
+{-# NOINLINE readInt #-}
+readInt = _readInt q
+  where q =
+# if WORD_SIZE_IN_BITS == 64
+            0xccccccccccccccc -- 0xccc_cccc_cccc_cccc
+# elif WORD_SIZE_IN_BITS == 32
+            0xccccccc         -- 0xccc_cccc
+# endif
+-- [Note @maxBound `quotRem` 10 == (q, 7)@]
+--
+-- It is not difficult to see that division of @2^n-1@ by @10@ produces a
+-- quotient that is a truncation of the infinite bitstring @1100_1100_...@.
+--
+-- When the input has 4n+3 bits, as is the case with 'maxBound' for all the
+-- Int* types, the quotient will have 4n bits with all the nibbles equal to
+-- @0xC@, and the remainder will be 7.
```

Comment:
I find this a little bit difficult to digest. I think there is a shorter proof available: 2^(4n+3)-1 = 8 * 16^n - 1 = 8 * ...6 - 1 = ...7, because 16^n ends in 6 for positive n.

Reply:
The idea was to also justify the `0xCCC...` quotient as well as the remainder.  The reason is that `1/5` in floating point binary is `0.001100110011...`  as is easily seen by shifting it left by 2 bits and adding (* 5).  Then `(2^n-1)/10` is just a left of this by `n-1` bits and drop the fractional part.  But if I go back to computing the quotient, that  becomes unnecessary to note...

Reply:
Getting rid of the 0xCCC magic in a shortly upcoming fixup...